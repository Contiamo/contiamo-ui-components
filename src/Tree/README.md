The tree component renders a tree structure with collapsable nodes in a filetree-like design.

### Usage

```jsx
import * as React from "react"
import { Tree } from "@operational/components"
;<Tree
  trees={[
    {
      label: "Store",
      childNodes: [
        {
          label: "Region",
          initiallyOpen: true,
          childNodes: [
            {
              label: "City",
              tag: "D",
              disabled: true,
              childNodes: [],
            },
            {
              label: "Country",
              color: "primary",
              onClick: () => alert("country was clicked"),
              onRemove: () => alert("node is removed"),
              tag: "D",
              childNodes: [],
            },
          ],
        },
      ],
    },
    {
      label: "Legal Entity",
      initiallyOpen: true,
      childNodes: [
        {
          label: "Limited Liability Company",
          tag: "D",
          childNodes: [],
        },
        {
          label: "Inc.",
          tag: "D",
          color: "#2C363F",
          childNodes: [],
        },
      ],
    },
  ]}
/>
```

### With Highlighting

Certain nodes in the tree can be highlighted like so.

```jsx
import * as React from "react"
import { Tree } from "@operational/components"
;<Tree
  trees={[
    {
      label: "Store",
      highlight: true,
      childNodes: [
        {
          label: "Region",
          initiallyOpen: true,
          childNodes: [
            {
              label: "City",
              tag: "D",
              disabled: true,
              childNodes: [],
            },
            {
              label: "Country",
              color: "primary",
              onRemove: () => {},
              tag: "D",
              childNodes: [],
            },
          ],
        },
      ],
    },
    {
      label: "Legal Entity",
      initiallyOpen: true,
      childNodes: [
        {
          label: "Limited Liability Company",
          tag: "D",
          childNodes: [],
          highlight: true,
        },
        {
          label: "Inc.",
          tag: "D",
          color: "#2C363F",
          childNodes: [],
        },
      ],
    },
  ]}
/>
```

### With react-beautiful-dnd

This tree component have `Droppable` and `Draggable` implemented from [react-beautiful-dnd](https://github.com/atlassian/react-beautiful-dnd/).

To offer the most flexibility as possible in your drag & drop implementation, we are just spreading `draggableProps` and `droppableProps`.

```js
import * as React from "react"
import { Button, Card, Tree, Title } from "@operational/components"
import { DragDropContext } from "react-beautiful-dnd"

// a little function to help us with reordering the result
const reorder = (list, startIndex, endIndex) => {
  const result = Array.from(list)
  const [removed] = result.splice(startIndex, 1)
  result.splice(endIndex, 0, removed)

  return result
}

const getMessage = item => {
  switch (item) {
    case "Dough":
      return "Always a good start!"
    case "Jalape√±o":
      return "Hot! üå∂"
    case "Pineapples":
      return "Really?!"
    case "Cheese":
      return "Oh yeah! üòé"
    default:
      return "Yum!"
  }
}
const PlaceHolder = ({ isDraggingOver, draggingOverWith }) => (
  <div style={{ minHeight: 30, border: "1px dashed gray", padding: 8, borderRadius: 3 }}>
    {isDraggingOver ? getMessage(draggingOverWith) : "Drag your stuff here"}
  </div>
)

const PizzaMaker = () => {
  const [inventory, setInventory] = React.useState([
    "Cheese",
    "Pepperoni",
    "Dough",
    "Mozzarella",
    "Tomato sauce",
    "Flour",
    "Pineapples",
    "Jalape√±o",
    "Artichokes",
    "Cauliflower",
    "Crushed red pepper",
    "Blue cheese",
    "Garlic",
    "Olive Oil",
    "Beef",
    "Pork",
  ])
  const [ingredients, setIngredients] = React.useState([])

  const onDragEnd = React.useCallback(result => {
    if (result.destination) {
      switch (result.destination.droppableId) {
        case "inventory":
          if (result.source.droppableId === "inventory") {
            // reorder
            setInventory(prev => reorder(prev, result.source.index, result.destination.index))
          } else {
            // ingredients -> inventory
            setIngredients(prev => prev.filter(i => i !== result.draggableId))
            setInventory(prev => [
              ...prev.slice(0, result.destination.index),
              result.draggableId,
              ...prev.slice(result.destination.index),
            ])
          }
          break
        case "ingredients":
          if (result.source.droppableId === "ingredients") {
            // reorder
            setIngredients(prev => reorder(prev, result.source.index, result.destination.index))
          } else {
            // inventory -> ingredients
            setInventory(prev => prev.filter(i => i != result.draggableId))
            setIngredients(prev => [
              ...prev.slice(0, result.destination.index),
              result.draggableId,
              ...prev.slice(result.destination.index),
            ])
          }
          break
      }
    }
  }, [])

  const cook = React.useCallback(() => {
    if (ingredients.includes("Mozzarella")) {
      alert("What an amazing pizza!")
    } else if (ingredients.includes("Pineapples")) {
      alert("Why are you doing this to a pizza?!")
    } else if (ingredients.includes("Jalape√±o")) {
      alert("Tasty!")
    } else {
      alert("Not bad ^^")
    }
  })

  return (
    <DragDropContext onDragEnd={onDragEnd}>
      <Title>Let's make a pizza!</Title>
      <div style={{ display: "grid", gridTemplateColumns: "1fr 1fr", grgridGap: 8, marginTop: 8 }}>
        <Card title="Inventory">
          <Tree
            droppableProps={{ droppableId: "inventory" }}
            trees={inventory.map(item => ({
              key: item,
              label: item,
              draggableProps: { draggableId: item },
            }))}
            placeholder={PlaceHolder}
          />
        </Card>
        <Card title="Ingredients">
          <Tree
            droppableProps={{ droppableId: "ingredients" }}
            trees={ingredients.map(item => ({
              key: item,
              label: item,
              draggableProps: { draggableId: item },
            }))}
            placeholder={PlaceHolder}
          />
        </Card>
        <Button fullWidth color="primary" onClick={cook} disabled={ingredients.length === 0}>
          Cook! üçï
        </Button>
      </div>
    </DragDropContext>
  )
}
;<PizzaMaker />
```
