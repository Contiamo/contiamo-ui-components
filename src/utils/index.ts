import urlRegex from "url-regex"

export { default as deprecate } from "./deprecate"
export * from "./mixins"
export * from "./color"
export { default as wrapDefaultTheme } from "./wrap-default-theme"

export const isModifiedEvent = (event: any) => !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey)

export const isOutsideLink = (url: string) => urlRegex({ exact: true }).test(url)

/**
 * Detect if a key event originated from a cmd+enter
 */
export const isCmdEnter = (ev: React.KeyboardEvent<HTMLElement>) => {
  const crossBrowserSafeKeycode = ev.which || ev.keyCode
  const modifierKey = ev.ctrlKey || ev.metaKey
  return crossBrowserSafeKeycode === keyCodes.enter && modifierKey
}

/**
 * Return the initials in 2 letters from a full name.
 *
 * @param name
 */
export const getInitials = (name: string): string => {
  if (!name) {
    return ""
  }
  const fullInitials = name
    .split(" ")
    .map(([firstLetter]) => firstLetter.toUpperCase())
    .join("")

  const [firstInitial, , lastInitial] = fullInitials
  return fullInitials.length > 2 ? firstInitial + lastInitial : fullInitials
}

export const keyCodes = {
  backspace: 8,
  tab: 9,
  clear: 12,
  enter: 13,
  return: 13,
  capsLock: 20,
  esc: 27,
  space: 32,
  left: 37,
  up: 38,
  right: 39,
  down: 40,
  del: 46,
  ins: 45,
  pageup: 33,
  pagedown: 34,
  end: 35,
  home: 36,
}

/**
 * Make a type and all its (sub) properties optional.
 */
export type DeepPartial<T> = T extends any[] ? DeepPartialArray<T[number]> : T extends object ? DeepPartialObject<T> : T

interface DeepPartialArray<T> extends Array<DeepPartial<T>> {}

type DeepPartialObject<T> = { [P in keyof T]+?: DeepPartial<T[P]> }
