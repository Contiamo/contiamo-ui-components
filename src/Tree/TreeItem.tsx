import React, { useCallback } from "react"

import NameTag from "../NameTag/NameTag"
import { darken, lighten } from "../utils"
import styled from "../utils/styled"
import { ChevronRightIcon, ChevronDownIcon, IconComponentType } from "../Icon"
import Highlighter from "react-highlight-words"
import constants from "../utils/constants"

interface TreeItemProps {
  level: number
  highlight: boolean
  searchWords?: string[]
  hasChildren: boolean
  isOpen: boolean
  label: string
  tag?: string
  icon?: IconComponentType
  iconColor?: string
  color?: string
  cursor?: string
  onNodeClick?: (e: React.MouseEvent<HTMLDivElement>) => void
  onNodeContextMenu?: (e: React.MouseEvent<HTMLDivElement>) => void
  onMouseEnter?: (e: React.MouseEvent<HTMLDivElement>) => void
  onMouseLeave?: (e: React.MouseEvent<HTMLDivElement>) => void
  actions?: React.ReactNode
  hasIconOffset?: boolean
}

const Header = styled.div<{
  highlight: boolean
  onClick?: (e: React.MouseEvent<HTMLDivElement>) => void
  cursor?: string
  level: number
  hasIconOffset: boolean
}>`
  label: TreeItem;
  display: flex;
  position: relative;
  align-items: center;
  cursor: ${({ onClick, cursor }) => cursor || (onClick ? "pointer" : "inherit")};
  background: ${({ highlight, theme }) => (highlight ? theme.color.highlight : "none")};
  padding: ${({ theme }) => `${theme.space.base / 2}px ${theme.space.content}px`};
  padding-left: ${({ theme, level, hasIconOffset }) =>
    theme.space.content * (level + 1) + (hasIconOffset ? theme.space.content - theme.space.base : -theme.space.base)}px;
  margin: 0 -${({ theme }) => theme.space.medium}px;
  color: ${({ theme }) => theme.color.text.dark};

  :hover,
  .no-focus &:hover:focus {
    background: ${({ theme, highlight }) =>
      highlight ? darken(theme.color.highlight, 20) : theme.color.background.lighter};

    /* Show ActionsContainer on hover */
    div:last-of-type {
      opacity: 1;
    }
  }

  :focus {
    outline: none;
    color: ${({ theme }) => theme.color.primary};
    background: ${({ theme }) => lighten(theme.color.primary, 50)};

    /* Show ActionsContainer on hover */
    div:last-of-type {
      opacity: 1;
    }
  }
  .no-focus &:focus {
    color: ${({ theme }) => theme.color.text.dark};
    background: ${({ highlight, theme }) => (highlight ? theme.color.highlight : "none")};
    div:last-of-type {
      opacity: 0;
    }
  }
`

const NameTagStyled = styled(NameTag)<{ withIcon?: boolean }>`
  label: NameTagStyled;
  margin-right: ${({ theme }) => theme.space.base}px;
`

// These props are extracted to avoid useless re-render
const highlightStyle: React.CSSProperties = {
  color: constants.color.text.action,
  backgroundColor: "transparent",
  fontWeight: "bold",
}
const defaultSearch: string[] = []

const Label = styled.div<{ hasChildren: boolean }>`
  /* Split the label by caract properly and show the first line only */
  overflow-wrap: break-word;
  overflow: hidden;
  text-overflow: ellipsis;
  height: 16px;

  font-size: ${({ theme }) => theme.font.size.small}px;
  font-weight: ${({ theme, hasChildren }) => (hasChildren ? theme.font.weight.bold : theme.font.weight.medium)};
  flex: 1;
`

const ActionsContainer = styled.div<{ childrenCount: number }>`
  display: grid;
  opacity: 0;
  align-items: center;
  right: 16px;
  grid-template-columns: repeat(${({ childrenCount }) => childrenCount}, 1fr);
  grid-column-gap: 4px;
`

const TreeItem: React.SFC<TreeItemProps> = ({
  highlight,
  tag,
  icon,
  iconColor,
  label,
  color,
  onNodeClick,
  onNodeContextMenu,
  hasChildren,
  isOpen,
  level,
  cursor,
  actions,
  hasIconOffset,
  searchWords = defaultSearch,
  onMouseEnter,
  onMouseLeave,
}) => {
  const handleKeyDown = useCallback(
    e => {
      switch (e.key) {
        case "Enter":
          e.preventDefault()
          if (e.altKey && onNodeContextMenu) {
            onNodeContextMenu(e)
            return
          }

          if (onNodeClick) {
            onNodeClick(e)
          }
          return
        case " ":
        case "Space": // the platformâ„¢
          e.preventDefault()
          if (onNodeClick) {
            onNodeClick(e)
          }
          return
      }
    },
    [onNodeContextMenu, onNodeClick],
  )

  return (
    <Header
      level={level}
      hasIconOffset={Boolean(hasIconOffset)}
      onClick={onNodeClick}
      onContextMenu={onNodeContextMenu}
      onKeyDown={handleKeyDown}
      highlight={Boolean(highlight)}
      cursor={cursor}
      tabIndex={0} // TODO: tabIndex -1 for disabled items
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      {hasChildren &&
        React.createElement(isOpen ? ChevronDownIcon : ChevronRightIcon, {
          size: 11,
          left: true,
          color: "color.text.lighter",
          style: { marginRight: 4 },
        })}
      {tag && (
        <NameTagStyled condensed left color={color}>
          {tag}
        </NameTagStyled>
      )}
      {icon &&
        React.createElement(icon, {
          size: 12,
          color: iconColor || "color.text.lighter",
          style: { marginLeft: 0, marginRight: 4, flex: "0 0 11px" },
        })}
      <Label hasChildren={hasChildren}>
        <Highlighter textToHighlight={label} highlightStyle={highlightStyle} searchWords={searchWords} />
      </Label>
      <ActionsContainer childrenCount={React.Children.count(actions)}>{actions}</ActionsContainer>
    </Header>
  )
}

export default TreeItem
