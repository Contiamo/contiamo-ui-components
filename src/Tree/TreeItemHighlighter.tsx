import React, { useMemo } from "react"
import Highlighter, { HighlighterProps } from "react-highlight-words"
import { withTheme } from "emotion-theming"

import { OperationalStyleConstants } from "../utils/constants"

const TreeItemHighlighter = ({ theme, highlightStyle, ...props }: HighlighterProps & OperationalStyleConstants) => {
  const computedHighlightStyle = useMemo(() => {
    return highlightStyle
      ? highlightStyle
      : {
          color: theme.color.text.action,
          backgroundColor: "transparent",
        }
  }, [highlightStyle, theme])

  return <Highlighter {...props} highlightStyle={computedHighlightStyle} />
}

export default withTheme(TreeItemHighlighter)
