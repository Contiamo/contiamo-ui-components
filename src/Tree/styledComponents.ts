import Icon from "../Icon/Icon"
import { darken } from "../utils"
import styled from "../utils/styled"

export const Header = styled("div")<{ highlight: boolean }>`
  display: flex;
  align-items: center;
  margin-top: 0;
  margin-bottom: ${({ theme }) => theme.space.base}px;
  cursor: pointer;
  background-color: ${({ highlight, theme }) => (highlight ? theme.color.highlight : "none")};
  outline: ${({ theme, highlight }) => (highlight ? `${theme.space.base}px solid ${theme.color.highlight}` : "none")};

  :hover {
    background-color: ${({ theme, highlight }) =>
      highlight ? darken(theme.color.highlight, 20) : theme.color.background.lighter};
    outline: ${({ theme, highlight }) =>
      `${theme.space.base}px solid ${highlight ? darken(theme.color.highlight, 20) : theme.color.background.lighter};`};
  }
`

export const Label = styled("div")<{ hasChildren: boolean }>`
  overflow-wrap: break-word;
  font-size: ${({ theme }) => theme.font.size.fineprint}px;
  font-weight: ${({ theme, hasChildren }) => (hasChildren ? theme.font.weight.bold : theme.font.weight.medium)};
  color: ${({ theme }) => theme.color.text.dark};
`

export const Container = styled("div")<{ hasChildren: boolean; disabled: boolean }>`
  opacity: ${({ disabled }) => (disabled ? 0.5 : 1)};
  pointer-events: ${({ disabled }) => (disabled ? "none" : "all")};
  user-select: none;

  & + & {
    margin-top: ${({ theme, hasChildren }) => (hasChildren ? theme.space.small : theme.space.base)}px;
  }
`

export const DeleteNode = styled("div")`
  display: flex;
  align-items: center;
  justify-content: center;
  margin-left: auto;
  width: 14px;
  height: 14px;
  background-color: rgba(0, 0, 0, 0.05);
  color: ${({ theme }) => theme.color.text.lighter};
  cursor: pointer;

  :hover {
    background-color: rgba(0, 0, 0, 0.1);
    cursor: pointer;
  }
`

export const TreeIcon = styled(Icon)`
  margin-right: ${({ theme }) => theme.space.base}px;
`
