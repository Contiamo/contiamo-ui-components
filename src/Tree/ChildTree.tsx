import * as React from "react"

import Icon from "../Icon/Icon"
import NameTag from "../NameTag/NameTag"
import { Container, DeleteNode, Header, Label, TreeIcon } from "./styledComponents"
import Tree, { TreeProps } from "./Tree"

type Props = TreeProps["trees"][-1]

const ChildTree: React.SFC<Props> = ({
  initiallyOpen,
  highlight,
  tag,
  label,
  color,
  disabled,
  forwardRef,
  childNodes = [],
  onClick: onNodeClick,
  onRemove,
  ...props
}) => {
  const [isOpen, setIsOpen] = React.useState(Boolean(initiallyOpen))
  const toggle = (e: React.MouseEvent<HTMLDivElement>) => {
    e.stopPropagation()
    if (childNodes && childNodes.length) {
      setIsOpen(!isOpen)
    }
  }

  const hasChildren = Boolean(childNodes && childNodes.length)

  return (
    <Container ref={forwardRef} disabled={Boolean(disabled)} hasChildren={hasChildren} {...props}>
      <Header
        onClick={e => {
          toggle(e)
          if (onNodeClick) {
            onNodeClick()
          }
        }}
        highlight={Boolean(highlight)}
      >
        {hasChildren && <TreeIcon color="color.text.lightest" size={12} left name={isOpen ? "ChevronDown" : "Add"} />}
        {!hasChildren && tag && (
          <NameTag condensed left color={color}>
            {tag}
          </NameTag>
        )}
        <Label hasChildren={hasChildren}>{label}</Label>
        {onRemove && (
          <DeleteNode onClick={onRemove}>
            <Icon size={12} name="No" />
          </DeleteNode>
        )}
      </Header>
      {childNodes && isOpen && <Tree trees={childNodes} />}
    </Container>
  )
}

export default ChildTree
