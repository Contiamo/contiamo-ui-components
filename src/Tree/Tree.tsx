import { Omit } from "emotion-theming/types/helper"
import * as React from "react"
import { Draggable, DraggableProps, Droppable, DroppableProps, DroppableStateSnapshot } from "react-beautiful-dnd"
import styled from "../utils/styled"
import ChildTree from "./ChildTree"
import { IconComponentType } from "../Icon"

interface BaseTree {
  label: string
  key?: string | number
  paddingLeft?: number
  paddingRight?: number
  highlight?: boolean
  initiallyOpen?: boolean
  tag?: string
  tagColor?: string
  disabled?: boolean
  icon?: IconComponentType
  iconColor?: string
  onClick?: (event: React.MouseEvent) => void
  onContextMenu?: (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void
  cursor?: string
  actions?: React.ReactNode
  forwardRef?: (element?: HTMLElement | null) => any
  strong?: boolean
  fontSize?: number
  fontColor?: string
  emphasized?: boolean
  monospace?: boolean
  ignoreSearchWords?: boolean
}

interface TreeWithChildren extends BaseTree {
  childNodes?: Tree[]
  draggableProps?: never
  droppableProps?: Omit<DroppableProps, "children">
}

interface TreeWithoutChildren extends BaseTree {
  childNodes?: never
  draggableProps?: Omit<DraggableProps, "children" | "index">
  droppableProps?: never
}

export type Tree = TreeWithChildren | TreeWithoutChildren

export interface TreeProps {
  trees: Tree[]
  paddingLeft?: number
  paddingRight?: number
  searchWords?: string[]
  droppableProps?: Omit<DroppableProps, "children">
  placeholder?: React.ComponentType<DroppableStateSnapshot>
  freeze?: boolean
  _level?: number
  _hasIconOffset?: boolean
}

const Container = styled.div`
  label: TreeContainer;
  user-select: none;
`

const Tree: React.SFC<TreeProps> = ({
  _level = 0,
  _hasIconOffset = false,
  paddingLeft,
  paddingRight,
  trees,
  droppableProps,
  placeholder,
  searchWords,
  freeze,
}) => {
  const isLowestLevel = trees.length === 0 || trees.some(tree => !tree.childNodes || !tree.childNodes.length)

  /**
   * If this is a category with children, no drag and drop
   * because only children can be dragged/sorted.
   */
  if (!isLowestLevel || !droppableProps) {
    return (
      <Container>
        {trees.map((treeData, index) => (
          <ChildTree
            paddingLeft={paddingLeft}
            paddingRight={paddingRight}
            level={_level}
            hasIconOffset={_hasIconOffset}
            key={index}
            {...treeData}
            searchWords={searchWords}
            freeze={freeze}
          />
        ))}
      </Container>
    )
  }

  return (
    <Droppable {...droppableProps}>
      {(droppableProvided, droppableSnapshot) => (
        <Container ref={droppableProvided.innerRef} {...droppableProvided.droppableProps}>
          {trees.length ? (
            <>
              {trees.map((treeData, index) => (
                <Draggable
                  key={treeData.key !== undefined ? treeData.key : index}
                  {...treeData.draggableProps || { draggableId: treeData.label }}
                  index={index}
                >
                  {draggableProvided => {
                    return (
                      <ChildTree
                        paddingLeft={paddingLeft}
                        paddingRight={paddingRight}
                        hasIconOffset={_hasIconOffset}
                        level={_level}
                        forwardRef={draggableProvided.innerRef}
                        searchWords={searchWords}
                        freeze={freeze}
                        {...treeData}
                        {...draggableProvided.draggableProps}
                        {...draggableProvided.dragHandleProps}
                      />
                    )
                  }}
                </Draggable>
              ))}
            </>
          ) : (
            placeholder && React.createElement(placeholder, droppableSnapshot)
          )}
          {droppableProvided.placeholder}
        </Container>
      )}
    </Droppable>
  )
}

export default Tree
