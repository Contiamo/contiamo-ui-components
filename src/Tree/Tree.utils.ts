import { Tree } from "./Tree.types"

export const arePathsEqual = (path1: number[], path2: number[]) => path1.join("-") === path2.join("-")

/**
 * Returns the largest element of an array of numbers, or undefined if the array is empty.
 */
const getMaxFromList = (nos: number[]): number | undefined => {
  if (nos.length === 0) {
    return undefined
  }
  return nos.reduce((accumulator, current) => (accumulator < current ? current : accumulator), -100000)
}

export const getDepth = (tree: Tree): number => {
  if (tree.childNodes.length === 0) {
    return 1
  }
  // Type-casting is necessary because at this point in the function there is a guarantee
  // that tree.childNodes is not empty and therefore it has a maximum value.
  return 1 + (getMaxFromList(tree.childNodes.map(getDepth)) as number)
}

export const getMaxDepth = (trees: Tree[]): number => getMaxFromList(trees.map(getDepth)) || 0

export const getInitialOpenPaths = (basePath: number[]) => (tree: Tree): number[][] => {
  return [
    ...(tree.initiallyOpen ? [basePath] : []),
    ...tree.childNodes.map((childTree, index) =>
      getInitialOpenPaths([...basePath, index])(childTree).reduce(
        (paths, accumulator) => [...paths, ...accumulator],
        [],
      ),
    ),
  ]
}

/**
 * Adds or removes a path from a list of paths.
 * togglePath([1, 2])([[1, 2, 3], [0]]) -> [[1, 2, 3], [0], [1, 2]]
 * togglePath([1, 2])([[1, 2], [0]]) -> [[0]]
 */
export const togglePath = (path: number[]) => (paths: number[][]): number[][] => {
  if (paths.length === 0) {
    return [path]
  }
  const [head, ...tail] = paths
  if (arePathsEqual(head, path)) {
    return tail
  }
  return [head, ...togglePath(path)(tail)]
}

export const containsPath = (path: number[]) => (paths: number[][]): boolean => {
  if (paths.length === 0) {
    return false
  }
  const [head, ...tail] = paths
  if (arePathsEqual(head, path)) {
    return true
  }
  return containsPath(path)(tail)
}
