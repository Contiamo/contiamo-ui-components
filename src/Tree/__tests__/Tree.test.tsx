import { render } from "enzyme"
import * as React from "react"
import { Tree as ThemelessTree } from "../../index"
import wrapDefaultTheme from "../../utils/wrap-default-theme"
import { getInitialOpenPaths, togglePath } from "../Tree.utils"

const Tree = wrapDefaultTheme(ThemelessTree)

describe("Tree Component", () => {
  it("Should intialize without problems", () => {
    const output = render(<Tree trees={[]} />)
    expect(output).toMatchSnapshot()
  })
  it("Toggle path - add case", () => {
    expect(togglePath([1, 2])([[1, 2, 3], [0]])).toEqual([[1, 2, 3], [0], [1, 2]])
  })
  it("Toggle path - remove case", () => {
    expect(togglePath([1, 2])([[1, 2], [0]])).toEqual([[0]])
  })
  it("Works out initially open paths", () => {
    expect(
      getInitialOpenPaths([])({
        label: "",
        initiallyOpen: true,
        childNodes: [
          {
            label: "",
            initiallyOpen: false,
            childNodes: [
              {
                label: "",
                initiallyOpen: true,
                childNodes: [],
              },
            ],
          },
          {
            label: "",
            initiallyOpen: true,
            childNodes: [],
          },
        ],
      }),
    ).toEqual([[], [0, 0], [1]])
  })
})
