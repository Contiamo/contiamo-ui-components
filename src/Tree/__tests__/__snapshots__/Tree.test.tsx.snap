// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree Component Should intialize without problems 1`] = `
<div
  class="css-ftf61z"
>
  <div
    class="css-ar73w8-Messages"
  />
  <div
    class="css-1i82ydo"
  />
</div>
`;
